### PR TITLE
[nnx] Add NNX WeightNorm.

### PR DESCRIPTION
**Pull request by nilq**
# What does this PR do?

This PR addresses #4426 by implementing an NNX WeightNorm class. This brings the requested normalization class to the NNX library – implemented to be consistent with the corresponding module in Linen.

## Checklist
- [x] This change is discussed in #4426.
- [x] The documentation and docstrings adhere to the
      [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [x] This change includes necessary high-coverage tests.

### CODE REVIEW

File: flax/nnx/nn/normalization.py [hunk 1]
```diff
@@ -835,4 +853,144 @@ def __call__(self, x, *, mask: tp.Optional[jax.Array] = None):
...
+    Returns:
+      Output of the layer using l2-normalized weights.
+    """
+    state = nnx.state(self.layer_instance)
+
+    def apply_weightnorm(path, var_state):
+      path_str = '/'.join(str(k) for k in path)
+
+      if self.variable_filter:
+        for variable_name in self.variable_filter:
```
Comment (cgarciae) [inline flax/nnx/nn/normalization.py:950]: We should use the same Filter supported by `nnx.filterlib` which are a bit more general and used thought the library.

**Comment by cgarciae** [comment]
To resolve the pytest issues you will have to add some deprecation warning to the list here:
https://github.com/google/flax/blob/6af8fcb0bfac1eb078bdf5ba41a3899a71dcaf81/pyproject.toml#L188

If this is not from your PR we can fix it in `main` first.

**Comment by nilq** [comment]
@cgarciae Just a friendly ping – in case this might have drowned. 🙂

**Comment by HyperBlaze456** [comment]
Hello, I think the variable filter is not correctly working?
The code does not apply them to extract only the required variables to be normalized, and there is only case where no variable filter is passed.

**Comment by samanklesaria** [comment]
Hi @nilq - we'd like to finally land this PR on top of all the recent changes to nnx! We just merged support for SpectralNorm (#4623), and your WeightNorm code is quite similar. Do you think it might be possible to factor out some of the shared functionality?

**Comment by samanklesaria** [comment]
Closed by #5043
